pkg/server/model.go

```diff
@@ -0,0 +1,140 @@
+package server
+
+import (
+	"strings"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+)
+
+const (
+	BigQueryTimestampFormat = "2006-01-02 15:04:05.999999 UTC"
+)
+
+type BigQueryTableQueryBuilder struct {
+	client *bigquery.Client
+	table  *bigquery.Table
+
+	parameters map[string]bigquery.QueryParameter
+}
+
+func (qb *BigQueryTableQueryBuilder) WithClient(client *bigquery.Client) {
+	qb.client = client
+}
+
+func (qb *BigQueryTableQueryBuilder) WithTable(table *bigquery.Table) {
+	qb.table = table
+}
+
+func (qb *BigQueryTableQueryBuilder) WithLog(logID string) {
+	qb.parameters["logID"] = bigquery.QueryParameter{
+		Name:  "logID",
+		Value: logID,
+	}
+}
+
+func (qb *BigQueryTableQueryBuilder) WithEncryptionKey(encryptionKey string) {
+	qb.parameters["encryptionKey"] = bigquery.QueryParameter{
+		Name:  "encryptionKey",
+		Value: encryptionKey,
+	}
+}
+
+func (qb *BigQueryTableQueryBuilder) WithStartAt(time *time.Time) {
+	qb.parameters["startAt"] = bigquery.QueryParameter{
+		Name:  "startAt",
+		Value: time.Format(BigQueryTimestampFormat),
+	}
+}
+
+func (qb *BigQueryTableQueryBuilder) WithEndAt(time *time.Time) {
+	qb.parameters["endAt"] = bigquery.QueryParameter{
+		Name:  "endAt",
+		Value: time.Format(BigQueryTimestampFormat),
+	}
+}
+
+func (qb *BigQueryTableQueryBuilder) After(time *time.Time) {
+	qb.parameters["after"] = bigquery.QueryParameter{
+		Name:  "after",
+		Value: time.Format(BigQueryTimestampFormat),
+	}
+}
+
+func (qb *BigQueryTableQueryBuilder) Before(time *time.Time) {
+	qb.parameters["before"] = bigquery.QueryParameter{
+		Name:  "before",
+		Value: time.Format(BigQueryTimestampFormat),
+	}
+}
+
+func (qb *BigQueryTableQueryBuilder) Build() (*bigquery.Query, error) {
+	var sb strings.Builder
+
+	sb.WriteString("SELECT ")
+	if _, ok := qb.parameters["encryptionKey"]; ok {
+		sb.WriteString("aead.decrypt_bytes(FROM_BASE64(@encryptionKey), encrypted_payload, b'')")
+	} else {
+		sb.WriteString("encrypted_payload")
```

Comment:
It feels like this should probably just be an error, rather than returning possibly-encrypted data back to a client. What do you think? Maybe just make the log ID and corresponding key required in the builder constructor?